cpp/src/arrow/json/parser-test.cc

```diff
@@ -0,0 +1,250 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <cstdint>
+#include <iomanip>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include <gtest/gtest.h>
+#include <rapidjson/error/en.h>
+#include <rapidjson/reader.h>
+
+#include "arrow/ipc/json-simple.h"
+#include "arrow/json/options.h"
+#include "arrow/json/parser.h"
+#include "arrow/json/reader.h"
+#include "arrow/json/test-common.h"
+#include "arrow/status.h"
+#include "arrow/testing/util.h"
+#include "arrow/util/logging.h"
+#include "arrow/util/string_view.h"
+
+namespace arrow {
+
+// using ipc::internal::json::ArrayFromJSON;
+using util::string_view;
+
+namespace json {
+
+std::string scalars_only_src() {
+  return R"(
+    { "hello": 3.5, "world": false, "yo": "thing" }
+    { "hello": 3.2, "world": null }
+    { "hello": 3.4, "world": null, "yo": "\u5fcd" }
+    { "hello": 0.0, "world": true, "yo": null }
+  )";
+}
+
+std::string nested_src() {
+  return R"(
+    { "hello": 3.5, "world": false, "yo": "thing", "arr": [1, 2, 3], "nuf": {} }
+    { "hello": 3.2, "world": null, "arr": [2], "nuf": null }
+    { "hello": 3.4, "world": null, "yo": "\u5fcd", "arr": [], "nuf": { "ps": 78 } }
+    { "hello": 0.0, "world": true, "yo": null, "arr": null, "nuf": { "ps": 90 } }
+  )";
+}
+
+void AssertRawStructArraysEqual(const StructArray& expected, const StructArray& actual);
+
+void AssertRawArraysEqual(const Array& expected, const Array& actual) {
```

Comment:
I wasn't sure how to name this; I was trying to express the array produced by parsing, preconversion. These correspond to JSON kinds rather than the array type after conversion, and have [their own builders](https://github.com/bkietz/arrow/blob/ARROW-694-json-reader-WIP/cpp/src/arrow/json/parser.cc#L137-138) because I needed to support efficiently initializing them with leading nulls (for a newly discovered field)